setuptools/config/_apply_pyprojecttoml.py

```diff
@@ -181,16 +182,19 @@ def _long_description(
         dist._referenced_files.add(file)
 
 
-def _license(dist: Distribution, val: dict, root_dir: StrPath | None):
+def _license(dist: Distribution, val: str | dict, root_dir: StrPath | None):
     from setuptools.config import expand
 
-    if "file" in val:
-        # XXX: Is it completely safe to assume static?
-        value = expand.read_files([val["file"]], root_dir)
-        _set_config(dist, "license", _static.Str(value))
-        dist._referenced_files.add(val["file"])
+    if isinstance(val, str):
+        _set_config(dist, "license_expression", _static.Str(val))
     else:
-        _set_config(dist, "license", _static.Str(val["text"]))
+        if "file" in val:
+            # XXX: Is it completely safe to assume static?
+            value = expand.read_files([val["file"]], root_dir)
+            _set_config(dist, "license", _static.Str(value))
+            dist._referenced_files.add(val["file"])
+        else:
+            _set_config(dist, "license", _static.Str(val["text"]))
```

Comment:
I would prefer to flatten this nested `else{if, else}` into a `elif,else` as in the [suggested change](https://github.com/pypa/setuptools/compare/97e8f94e24a57a446dee965f71532acacf7354d9..f3a1fdddf62fc61195b9a5e2ed8c66f4b4613238#diff-3c50c435ad8328e83caa459b99929ae4273860772b44c14f49039927636c5c04L182):

![image](https://github.com/user-attachments/assets/23d0446e-8d49-453e-9496-41a909339eb3)

Reply:
I thought about this as well. However I decided to use to nesting to clearly separate the before / after PEP 639 cases. In a flattened list it just looks like they are equal which isn't the case.

At some point, it might also make sense to add a DeprecationWarning for the legacy case which will be easier with the nested style.